libompd/src/TargetValue.cpp

```diff
@@ -349,6 +349,60 @@ ompd_rc_t TValue::getRawValue(void *buf, int count) {
   return errorState.errorCode;
 }
 
+ompd_rc_t TValue::getString(const char **buf) {
+  if (gotError())
+    return getError();
+
+  TValue strValue = dereference();
+  if (strValue.gotError() && (strValue.getError() != ompd_rc_unsupported)) {
+    return strValue.getError();
+  }
+
+  ompd_rc_t ret;
+
+  if (strValue.gotError() && (strValue.getError() == ompd_rc_unsupported)) {
+    if (strValue.symbolAddr.address == 0) {
+      // Pass back an empty string if the pointer is NULL.
```

Comment:
Why? I'm not sure I understand the logic here. Seems to me that any error should result in setting the string pointer to null, that is: *buf = 0;

Reply:
As a part of my testing, I noticed that variables like "__kmp_tool_libraries" (representing the tool-libraries-var ICV) would be NULL, if the corresponding env variable "OMP_TOOL_LIBRARIES" is not set. This is a valid scenario and dereference() sets the error code to ompd_rc_unsupported (which I am not sure is the right approach -- I think this should be an ompd_rc_ok) in this case.  Anyways, hence, I am returning an empty string ("") with ompd_rc_ok in this case so that the tool can continue.  From what I could make out, the other scenarios where dereference() sets an error code, seemed to be genuine erroneous situations where probably the tool should not continue.

Reply:
Failed to mention that I will be placing a comment in the code explaining this.

Reply:
I think the getString() method you added should have semantics that are independent of the semantics of the ICV you are reading, and it should be possible to distinguish a null pointer from an empty string. The caller of getString() should impose the additional semantics of mapping a null pointer to an empty string, if appropriate for the ICV being fetched. The OMP v5 spec says that the initial value of "tool-libraries-var" is the empty string, so when fetching that ICV it might make sense to map the null to an empty string. Or, perhaps the OMP runtime is wrong and it should initialize __kmp_tool_libraries to the empty string.

Reply:
Ok, thanks, makes sense. I can have the callers of getString() allocate and return the null strings if needed. And __kmp_tool_libraries being NULL instead of pointing to an empty string seems like a runtime error to me too. @jprotze, do you also think so ? I can probably correct this in a future patch if so.

Reply:
> This is a valid scenario and dereference() sets the error code to ompd_rc_unsupported (which I am not sure is the right approach -- I think this should be an ompd_rc_ok) in this case.

The reason, why dereference sets an error code for dereferencing NULL is to allow for chaining the TargetValue methods. As soon as a TargetValue objects has an error, it will not apply any further methods but just pass on the error code. If we see need for more detailed localization of the source of such error, this can be implemented in the error class.

Reply:
Thank you, @jprotze. Do you think we should consider another (possibly new) error code in this case -- so that the callers can differentiate this scenario from others ? Currently, in my new commit, I am relying on ompd_rc_unsupported only for this.